### PR TITLE
Add UOps pattern matcher regression tests

### PR DESCRIPTION
**Pull request by alecchendev**
Standalone PR to add simple tests for each of the important fields that gets matched on.

### CODE REVIEW

File: test/test_pattern_matcher.py [hunk 1]
```diff
@@ -17,6 +17,56 @@ def test_simple_match(self):
     self.assertEqual(matcher.rewrite(c1), c1)
     self.assertEqual(matcher.rewrite(c2), None)
 
+  def test_uop(self):
+    matcher = PatternMatcher([({"__name__": "x", "uop": UOps.CONST}, lambda x: x)])
+    c1 = UOp(UOps.CONST)
```
Comment (Qazalin) [inline test/test_pattern_matcher.py:22]: please check the UOp spec https://github.com/tinygrad/tinygrad/blob/master/tinygrad/codegen/uops.py#L398 to create a correct UOp.

Comment (alecchendev) [reply test/test_pattern_matcher.py:22]: thanks, will fix this later today

Comment (alecchendev) [reply test/test_pattern_matcher.py]: done!

File: test/test_pattern_matcher.py [hunk 2]
```diff
@@ -29,12 +79,14 @@ def test_dtype_set(self):
     self.assertEqual(matcher.rewrite(c4), None)
 
   def test_vin_one(self):
-    matcher = PatternMatcher([({"__name__": "x", "uop": UOps.ALU, "vin":({"uop": UOps.CONST}, {"uop": UOps.CONST})}, lambda x: x)])
```
Comment (Qazalin) [inline test/test_pattern_matcher.py:32]: can you keep this and add a new one for (CONST, ALU)

File: test/test_pattern_matcher.py [hunk 3]
```diff
@@ -17,6 +17,56 @@ def test_simple_match(self):
...
+    c1 = UOp(UOps.CONST, dtypes.float, arg=1.0)
+    c2 = UOp(UOps.ALU, dtypes.float, (c1, c1), BinaryOps.ADD)
+    self.assertEqual(matcher.rewrite(c1), c1)
+    self.assertEqual(matcher.rewrite(c2), None)
+
+  def test_uop_set(self):
+    matcher = PatternMatcher([({"__name__": "x", "uop": {UOps.CONST, UOps.DEFINE_ACC}}, lambda x: x)])
+    c1 = UOp(UOps.CONST, dtypes.float, arg=1.0)
+    c2 = UOp(UOps.ALU, dtypes.float, (c1, c1), BinaryOps.ADD)
+    c3 = UOp(UOps.DEFINE_ACC, dtypes.int, arg=(1,))
```
Comment (Qazalin) [inline test/test_pattern_matcher.py:31]: did you check how many args DEFINE_ACC has?
https://github.com/tinygrad/tinygrad/blob/master/tinygrad/codegen/linearizer.py#L106t

**Comment by Qazalin** [comment]
I still 2+ wrong UOps.

File: test/test_pattern_matcher.py (hunk 3 above)
Comment (alecchendev) [reply test/test_pattern_matcher.py]: i initially just ran all my defined UOps through the `type_verify` function and assumed that would be enough, my mistake. just switched to use mainly CONST and ALU ops to simplify things

File: test/test_pattern_matcher.py [hunk 4]
```diff
@@ -17,6 +17,56 @@ def test_simple_match(self):
...
+    c1 = UOp(UOps.CONST, dtypes.float, arg=1.0)
+    c2 = UOp(UOps.ALU, dtypes.float, (c1, c1), BinaryOps.ADD)
+    self.assertEqual(matcher.rewrite(c1), c1)
+    self.assertEqual(matcher.rewrite(c2), None)
+
+  def test_uop_set(self):
+    matcher = PatternMatcher([({"__name__": "x", "uop": {UOps.CONST, UOps.CAST}}, lambda x: x)])
+    c1 = UOp(UOps.CONST, dtypes.bool, arg=False)
+    c2 = UOp(UOps.ALU, dtypes.float, (c1, c1), BinaryOps.ADD)
+    c3 = UOp(UOps.CAST, dtypes.int, (c1,))
```
Comment (Qazalin) [inline test/test_pattern_matcher.py:31]: wrong

Comment (alecchendev) [reply test/test_pattern_matcher.py]: 🤦 alright, i think this should be the last one

File: test/test_pattern_matcher.py [hunk 5]
```diff
@@ -17,6 +17,57 @@ def test_simple_match(self):
...
+    matcher = PatternMatcher([({"__name__": "x", "uop": UOps.CONST}, lambda x: x)])
+    c1 = UOp(UOps.CONST, dtypes.float, arg=1.0)
+    c2 = UOp(UOps.ALU, dtypes.float, (c1, c1), BinaryOps.ADD)
+    self.assertEqual(matcher.rewrite(c1), c1)
+    self.assertEqual(matcher.rewrite(c2), None)
+
+  def test_uop_set(self):
+    matcher = PatternMatcher([({"__name__": "x", "uop": {UOps.CONST, UOps.CAST}}, lambda x: x)])
+    c1 = UOp(UOps.CONST, dtypes.bool, arg=False)
+    c2 = UOp(UOps.CAST, dtypes.int, (c1,))
```
Comment (Qazalin) [inline test/test_pattern_matcher.py:30]: wrong

**Comment by geohot** [comment]
@Qazalin feel free to merge if this is correct and useful

**Comment by Qazalin** [comment]
`test_dup_name` is testing an invalid graph since y1 and y2 cannot exist in the same UOpGraph, I'd keep it though it's a useful behavior test.

**Pull request merged**